### PR TITLE
Fix #1873 for classpath snapshotting

### LINKED ISSUE #1873: Resources with identical content but different names causes tasks using `@Classpath` to be out-of-date
### Expected Behavior

<3.5 behavior.  Visiting resources with identical content in different orders should not generate different cache keys.

### Current Behavior

When a task uses the `@Classpath` annotation, Gradle uses a specialized snapshotter that ignores file order, jar content timestamps (and order) and some other things. We cache the result of snapshotting each file by keying off the MD5 hash of the _content_ of the file.  This is fine if the contents of the file is the only thing that goes into calculating the snapshot/signature.

I believe this can occur on a single machine when switching between different GRADLE_USER_HOME (or clearing Gradle's internal caches), but it's most visible when using a remote build cache.

### Steps to Reproduce (for bugs)

Reproducing this is difficult.  This is what you need to have to reproduce:

- Resources on the classpath (anything other than a .jar or .class) need to have different names, but identical contents
- Resources need to appear in different orders across machines.  This usually means they have to be in the same directory (or directory hierarchy).  It would be difficult to make this happen with files that are in different projects.
- The sequence is:
   - You have resources A and B. md5(A) == md5(B).
   - Gradle sees A first and records hashForA for the md5(A) key.  hashForA is a combination of md5(A) + A's name.
   - Gradle then sees B. It reuses hashForA, since the file has the same md5. This is wrong.
   - The cache key is created by using 2*hashForA.
   - On another machine...Gradle sees B first and records hashForB for the md5(B) key.  hashForB is a combination of md5(B) + B's name.
   - Gradle then sees B. It reuses hashForA, since the file has the same md5. This is wrong.
   - The cache key is created by using 2*hashForB.
   - hashForA != hashForB, so the second machine would experience a cache miss.

### Your Environment

This becomes an big issue when using the build cache, since the file-visiting order is OS dependent.  macOS-only builds would probably never see a problem (since the ordering seems to be consistent).

### PR DESCRIPTION
**Pull request by wolfs**
This PR fixes #1873 by consistently using the right normalization and compare strategies for jar file contents and directory contents. #1873 was introduced by the fix for #1752. We use the same strategy for the compile classpath and the runtime classpath.

### CODE REVIEW

File: subprojects/core/src/integTest/groovy/org/gradle/api/tasks/CachedTaskExecutionIntegrationTest.groovy [hunk 1]
```diff
@@ -438,7 +436,7 @@ class CachedTaskExecutionIntegrationTest extends AbstractIntegrationSpec impleme
         withBuildCache().succeeds("cacheable")
 
         when:
-        gradleUserHome.deleteDir() // nuke the cache
+        gradleUserHome.deleteDir() // nuke the file snapshot cache
```
Comment (adammurdoch) [inline subprojects/core/src/integTest/groovy/org/gradle/api/tasks/CachedTaskExecutionIntegrationTest.groovy]: Does this actually work on the windows daemon test suite?

File: subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/JarContentHasher.java [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+        this.contentHasher = contentHasher;
+        this.stringInterner = stringInterner;
+    }
+
+    public HashCode hash(RegularFileSnapshot jarFile) {
+        String jarFilePath = jarFile.getPath();
+        ZipInputStream zipInput = null;
+        try {
+            ClasspathEntrySnapshotBuilder entryResourceCollectionBuilder = new ClasspathEntrySnapshotBuilder(stringInterner);
+            zipInput = new ZipInputStream(Files.newInputStream(Paths.get(jarFilePath)));
```
Comment (adammurdoch) [inline subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/JarContentHasher.java]: This won't close the if the zip file is badly formed

Comment (wolfs) [inline subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/JarContentHasher.java]: Good catch. I fixed this.
Interestingly, `ZipInputStream` does not fail on an invalid zip file - for example a text file - but returns an empty set of entries. Should we try to detect this case? Or do you know a better implementation for treating zip files?

File: subprojects/core/src/integTest/groovy/org/gradle/api/tasks/CachedTaskExecutionIntegrationTest.groovy (hunk 1 above)
Comment (wolfs) [inline subprojects/core/src/integTest/groovy/org/gradle/api/tasks/CachedTaskExecutionIntegrationTest.groovy]: Seems to work: https://builds.gradle.org/viewLog.html?buildTypeId=Gradle_Branches_CoveragePhase_WindowsCoverage_WindowsJava17DaemonIntegrationTe_7&buildId=2815424

**Pull request merged**